### PR TITLE
[TOPI] Add conv2d int8 template

### PR DESCRIPTION
**Pull request by vinx13**
This PR added a int8 conv2d using NCHW[x]c layout, where x is multiple of 4. I obtained best performance when x = 4.
The template can accept either NCHW layout input or pre-packed data (NCHW4c) and kernel (OIHW4o4i). 

inference time (ms) of different models (before classifier) on NVIDIA 1080, batch size = 1

| model              | TVM (int8) | TVM (fp32) | TensorRT (int8)  | mxnet + cuDNN (fp32) |
|-------------------|----------------|----------------|----------------------|-----|
| vgg-16            |  1.64           | 4.13            | 1.44                  | 5.14 |
| resnet-50        |  1.46           | 3.95            | 1.39                  | 5.38 |
| inception_v3   | 2.55            |  7.98           |  2.30                 | 12.51 |

cc @merrymercy @tqchen

### CODE REVIEW

**Comment by tqchen** [comment]
@masahi @nishi-t please help review this as well

File: topi/python/topi/cuda/conv2d_int8.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+
+    ic_block_factor = 4
+    oc_block_factor = 4
+
+    if not pre_computed:
+        batch, channels, height, width = get_const_tuple(data.shape)
+        assert channels % ic_block_factor == 0, \
+            "Number of input channels should be multiple of {}".format(
+                ic_block_factor)
+        packed_data = tvm.compute((batch, channels/ic_block_factor, height, width, ic_block_factor),
```
Comment (masahi) [inline topi/python/topi/cuda/conv2d_int8.py]: channel //

File: topi/python/topi/cuda/conv2d_int8.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+                                                                c*ic_block_factor + vc, h, w],
+                                  name="packed_data")
+
+        out_channels, in_channels, kernel_h, kernel_w = get_const_tuple(
+            kernel.shape)
+        assert out_channels % 4 == 0, \
+            "Number of output channels should be multiple of {}".format(
+                oc_block_factor)
+        packed_kernel = tvm.compute(
+            (out_channels / oc_block_factor, in_channels / ic_block_factor, kernel_h, kernel_w,
```
Comment (masahi) [inline topi/python/topi/cuda/conv2d_int8.py]: //

File: topi/python/topi/cuda/conv2d_int8.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+    # unroll
+    cfg.define_knob("auto_unroll_max_step", [0, 512, 1500])
+    s[output].pragma(kernel_scope, 'auto_unroll_max_step',
+                     cfg['auto_unroll_max_step'].val)
+    s[output].pragma(kernel_scope, 'unroll_explicit', False)
+
+    return s
+
+
+@conv2d_NCHWc_int8_prepacked.register(["cuda", "gpu"])
```
Comment (masahi) [inline topi/python/topi/cuda/conv2d_int8.py]: "gpu" should be removed, since you are using cuda specific intrinsic

File: topi/python/topi/cuda/conv2d_int8.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+    s: Schedule
+        The computation schedule for conv2d.
+    """
+    s = tvm.create_schedule([x.op for x in outs])
+
+    def _callback(op):
+        if 'conv2d_NCHWc_int8' in op.tag:
+            schedule_conv2d_NCHWc_int8(cfg, s, op.output(0), pre_computed=True)
+
+    traverse_inline(s, outs[0].op, _callback)
```
Comment (vinx13) [inline topi/python/topi/cuda/conv2d_int8.py]: With bias enabled, I encountered some direct host side access error because bias is not computed inline. @merrymercy Could you help?

Comment (masahi) [reply topi/python/topi/cuda/conv2d_int8.py]: you need to handle bias and other elemwise op (relu, batch norm, etc) inside conv2d schedule. In cuda schedule, they are handled [here](https://github.com/dmlc/tvm/blob/master/topi/python/topi/cuda/conv2d_direct.py#L42-L45).

File: topi/python/topi/cuda/conv2d_int8.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+    oc_block_factor = 4
+
+    if not pre_computed:
+        batch, channels, height, width = get_const_tuple(data.shape)
+        assert channels % ic_block_factor == 0, \
+            "Number of input channels should be multiple of {}".format(
+                ic_block_factor)
+        packed_data = tvm.compute((batch, channels // ic_block_factor, height, width,
+                                   ic_block_factor),
+                                  lambda n, c, h, w, vc: kernel[n,
```
Comment (FrozenGene) [inline topi/python/topi/cuda/conv2d_int8.py]: Should be data? not kerne?

File: topi/python/topi/cuda/conv2d_winograd.py [hunk 6]
```diff
@@ -375,6 +375,12 @@ def _alter_conv2d_layout(attrs, inputs, tinfos):
         if cfg.template_key == 'direct':
             return None
 
+        if cfg.template_key == 'int8':
+            new_attrs['layout'] = 'NCHW4c'
```
Comment (masahi) [inline topi/python/topi/cuda/conv2d_winograd.py]: _alter_conv2d_layout is called from opencl and rocm backend too. They don't support int8, so it is better to add an assertion here.
 
```
assert tvm.target.current_target() in ["cuda", "nvptx"]
``` 

I'm not sure if dp4 intrinsic works on nvptx backend. If not, you can remove "nvptx" from above.

File: topi/python/topi/cuda/conv2d_winograd.py [hunk 7]
```diff
@@ -375,6 +375,12 @@ def _alter_conv2d_layout(attrs, inputs, tinfos):
...
+            new_attrs['out_layout'] = 'NCHW4c'
+            new_attrs['kernel_layout'] = 'OIHW4o4i'
+            return sym.contrib.conv2d_NCHWc_int8_prepacked(*copy_inputs, **new_attrs)
```
Comment (merrymercy) [inline topi/python/topi/cuda/conv2d_winograd.py]: Can we just return sym.conv2d?
I think this new symbol is redundant.

We need new symbol for winograd because we have to use a different infer_shape. But for conv2d, the infer_shape in nnvm already supports these layouts.

Comment (FrozenGene) [reply topi/python/topi/cuda/conv2d_winograd.py]: conv2d currently only accept NCHW / NHWC. If we pass NCHW4c, currently will meet layout assert error.

Comment (merrymercy) [reply topi/python/topi/cuda/conv2d_winograd.py]: Do you mean this ?
 https://github.com/dmlc/tvm/blob/ec0d497c69ca307fb998c3d81c0a7e48bb5f18d6/nnvm/python/nnvm/top/nn.py#L140-L146

It's better to change this instead of defining new op

Comment (merrymercy) [reply topi/python/topi/cuda/conv2d_winograd.py]: @vinx13  Yes, the problem is the one @FrozenGene just pointed out. I think we can fix it.

File: topi/python/topi/cuda/conv2d_int8.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+    kernel_scope, n = s[output].split(n, nparts=1)
+
+    max_block_z = 128
+    if batch > max_block_z:
+        _, n = s[output].split(n, factor=max_block_z)
+    s[output].bind(n, tvm.thread_axis("blockIdx.z"))
+
+    s[output].bind(bf, tvm.thread_axis("blockIdx.y"))
+    s[output].bind(bx, tvm.thread_axis("blockIdx.x"))
+    s[output].bind(vf, tvm.thread_axis("vthread"))
```
Comment (merrymercy) [inline topi/python/topi/cuda/conv2d_int8.py]: We bind `n` but don't bind or fuse `by`. Can you explain why you choose this strategy for batch?

Comment (vinx13) [reply topi/python/topi/cuda/conv2d_int8.py]: binding `n` can be very effective when batch size is large. I tested on some cases that fusing `by` can be slower, but I guess it should be tuneable.

File: topi/python/topi/cuda/conv2d_winograd.py (hunk 7 above)
Comment (vinx13) [reply topi/python/topi/cuda/conv2d_winograd.py]: @merrymercy another problem is that we can't call a topi template directly with packed layout, instead I registered a workload function to create original workload (in NCHW layout) from packed input. 
Adding this works.
https://github.com/dmlc/tvm/blob/ecad8bf05e80804218f8ef02bbc5c4337d247783/nnvm/python/nnvm/top/nn.py#L108

File: nnvm/python/nnvm/top/nn.py [hunk 9]
```diff
@@ -136,6 +143,8 @@ def schedule_conv2d(attrs, outs, target):
     with tvm.target.create(target):
         if groups == 1 and layout == "NCHW":
             return topi.generic.schedule_conv2d_nchw(outs)
+        elif groups == 1 and layout == "NCHW4c":
+            return topi.generic.schedule_conv2d_NCHWc_int8_prepacked(outs)
```
Comment (vinx13) [inline nnvm/python/nnvm/top/nn.py]: @merrymercy I can't check dtype of input here. Could you comment here?

**Comment by vinx13** [comment]
Currently, we cannot run a `conv2d+bn` model directly because of error with int8 bn weights.
In `sym.batch_norm(data=int8_conv)`, error occurs when it tries to compute sqrt of int in `fuse___add_scalar___sqrt___rdiv_scalar___elemwise_mul` because bn params are int8.
Some hack may be needed to compute bn params in fp32.
@tqchen Could you help?

**Comment by tqchen** [comment]
@masahi @merrymercy @FrozenGene can you followup on this and https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

File: topi/python/topi/cuda/conv2d_int8.py [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+
+        fused, tx = s[load].split(fused, factor=cfg["tile_x"].size[2])
+        fused, ty = s[load].split(fused, factor=cfg["tile_y"].size[2])
+        fused, tz = s[load].split(fused, factor=cfg["tile_f"].size[2])
+        s[load].bind(tz, tvm.thread_axis("threadIdx.z"))
+        s[load].bind(ty, tvm.thread_axis("threadIdx.y"))
+        s[load].bind(tx, tvm.thread_axis("threadIdx.x"))
+
+    # double buffer
+    cfg.define_knob('AA_double_buffer', [0, 1])
```
Comment (masahi) [inline topi/python/topi/cuda/conv2d_int8.py]: Just curious, are there cases where double buffering helped?

Comment (vinx13) [reply topi/python/topi/cuda/conv2d_int8.py]: I checked tuned result of AutoTVM. In vgg, most layers use double buffering. But in resnet, it seems this doesn't help. Some layers use double buffering, while others do not.

**Comment by FrozenGene** [comment]
Just one question, could we not set NCHW(x)c to NCHW4c explicitly but give it to AutoTVM decide how to split input channel?

**Comment by vinx13** [comment]
@FrozenGene yes we could let AutoTVM to tune the input channel split factor. But this will need some change in `alter_conv2d_layout` and some extra layout transform will be needed. I think this will be a good idea when we have graph tuner. Actually, I have tried x = 4/8/16 on all resnet layers, and found that perf of 4 > 8 > 16.

File: nnvm/python/nnvm/top/nn.py (hunk 9 above)
Comment (merrymercy) [reply nnvm/python/nnvm/top/nn.py]: Can we get dtype from outs[0].dtype?

Comment (vinx13) [reply nnvm/python/nnvm/top/nn.py]: @merrymercy outs[0].dtype is a user input, which can be arbitrary

**Comment by tqchen** [comment]
Thanks @merrymercy  @FrozenGene @nishi-t for review and @vinx13 for contribution, this is merged. Let us follow up to see if we can generalize the layout changes
